sys/suit/coap.c

```diff
@@ -0,0 +1,514 @@
+#endif
+
+#ifndef SUIT_MANIFEST_BUFSIZE
+#define SUIT_MANIFEST_BUFSIZE   640
+#endif
+
+#define SUIT_MSG_TRIGGER        0x12345
+
+static char _stack[SUIT_COAP_STACKSIZE];
+static char _url[SUIT_URL_MAX];
+static uint8_t _manifest_buf[SUIT_MANIFEST_BUFSIZE];
+
+static kernel_pid_t _suit_coap_pid;
+
+ssize_t coap_subtree_handler(coap_pkt_t *pkt, uint8_t *buf, size_t len,
+                             void *context)
+{
+    uint8_t uri[NANOCOAP_URI_MAX];
+
+    unsigned method_flag = coap_method2flag(coap_get_code_detail(pkt));
+
+    if (coap_get_uri_path(pkt, uri) > 0) {
+        coap_resource_subtree_t *subtree = context;
+
+        for (unsigned i = 0; i < subtree->resources_numof; i++) {
+            const coap_resource_t *resource = &subtree->resources[i];
+
+            if (!(resource->methods & method_flag)) {
+                continue;
+            }
+
+            int res = coap_match_path(resource, uri);
+            if (res > 0) {
+                continue;
+            }
+            else if (res < 0) {
+                break;
+            }
+            else {
+                return resource->handler(pkt, buf, len, resource->context);
+            }
+        }
+    }
+
+    return coap_reply_simple(pkt, COAP_CODE_INTERNAL_SERVER_ERROR, buf,
+                             len, COAP_FORMAT_TEXT, NULL, 0);
+}
+
+/* verbatim copy from sys/net/application_layer/nanocoap/nanocoap.c */
+static size_t _encode_uint(uint32_t *val)
+{
+    uint8_t *tgt = (uint8_t *)val;
+    size_t size = 0;
+
+    /* count number of used bytes */
+    uint32_t tmp = *val;
+    while(tmp) {
+        size++;
+        tmp >>= 8;
+    }
+
+    /* convert to network byte order */
+    tmp = htonl(*val);
+
+    /* copy bytewise, starting with first actually used byte */
+    *val = 0;
+    uint8_t *tmp_u8 = (uint8_t *)&tmp;
+    tmp_u8 += (4 - size);
+    for (unsigned n = 0; n < size; n++) {
+        *tgt++ = *tmp_u8++;
+    }
+
+    return size;
+}
+
+size_t coap_put_option_block(uint8_t *buf, uint16_t lastonum, unsigned blknum, unsigned szx, int more, uint16_t option)
+{
+    uint32_t blkopt = (blknum << 4) | szx | (more ? 0x8 : 0);
+    size_t olen = _encode_uint(&blkopt);
+
+    return coap_put_option(buf, lastonum, option, (uint8_t *)&blkopt, olen);
+}
+
+static ssize_t _nanocoap_request(sock_udp_t *sock, coap_pkt_t *pkt, size_t len)
+{
+    ssize_t res;
+    size_t pdu_len = (pkt->payload - (uint8_t *)pkt->hdr) + pkt->payload_len;
+    uint8_t *buf = (uint8_t*)pkt->hdr;
+
+    /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT *
+     * ACK_RANDOM_FACTOR) */
+    uint32_t timeout = COAP_ACK_TIMEOUT * US_PER_SEC;
+    unsigned tries_left = COAP_MAX_RETRANSMIT + 1;  /* add 1 for initial transmit */
+    while (tries_left) {
+
+        res = sock_udp_send(sock, buf, pdu_len, NULL);
+        if (res <= 0) {
+            DEBUG("nanocoap: error sending coap request, %d\n", (int)res);
+            break;
+        }
+
```

Comment:
Sometimes because of re-transmission and old block is accepted before the requested one:

```
2019-09-27 13:07:06,657 - INFO # riotboot_flashwrite: processing bytes 10240-10303
2019-09-27 13:07:06,697 - INFO # _suit_flashwrite(): writer->offset=10304, offset==10240, aborting
2019-09-27 13:07:06,699 - INFO # image download failed
2019-09-27 13:07:06,701 - INFO # )handler returned <0
2019-09-27 13:07:06,702 - INFO # )handler returned <0
2019-09-27 13:07:06,705 - INFO # )suit_v4_parse() failed. res=-1
```

This leads to un-needed failures I think we should check the id with a timeout, something like:

```
    uint32_t id = coap_get_id(pkt);
    ssize_t res = -1;

    while (tries_left && res) {
        while(xtimer_now() < block_timeout){
            res = sock_udp_recv(sock, buf, len, timeout, NULL);
            if (res <= 0) {
                if (res == -ETIMEDOUT) {
                    DEBUG("nanocoap: timeout\n");

                    timeout *= 2;
                    tries_left--;
                    if (!tries_left) {
                        DEBUG("nanocoap: maximum retries reached\n");
                    }
                    continue;
                }
                DEBUG("nanocoap: error receiving coap response, %d\n", (int)res);
                break;
            }
            else {
                if (coap_parse(pkt, (uint8_t *)buf, res) < 0) {
                    DEBUG("nanocoap: error parsing packet\n");
                    res = -EBADMSG;
                }
                if (coap_get_id(pkt) == id) {
                    res = 0;
                    break;
                }
            }
        }
    }
```

Or check the id when `_nanocoap_request()` is called (L199)

Reply:
I edited the code suggestion, I had removed the send part :) (still I haven't actually tested the code, consider it pseudocode)

Reply:
Here is the suggested change.

<details><summary>PATCH</summary>

```
diff --git a/sys/suit/coap.c b/sys/suit/coap.c
index 9e86777b3..870bb8d7f 100644
--- a/sys/suit/coap.c
+++ b/sys/suit/coap.c
@@ -141,15 +141,20 @@ size_t coap_put_option_block(uint8_t *buf, uint16_t lastonum, unsigned blknum, u
 
 static ssize_t _nanocoap_request(sock_udp_t *sock, coap_pkt_t *pkt, size_t len)
 {
-    ssize_t res;
+    ssize_t res = 0;
     size_t pdu_len = (pkt->payload - (uint8_t *)pkt->hdr) + pkt->payload_len;
     uint8_t *buf = (uint8_t*)pkt->hdr;
 
     /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT *
      * ACK_RANDOM_FACTOR) */
     uint32_t timeout = COAP_ACK_TIMEOUT * US_PER_SEC;
+    /* Timeout on incorrect block responses */
+    uint32_t block_timeout = timeout + xtimer_now_usec();
+    uint32_t id = coap_get_id(pkt);
+
     unsigned tries_left = COAP_MAX_RETRANSMIT + 1;  /* add 1 for initial transmit */
-    while (tries_left) {
+
+    while (tries_left && res) {
 
         res = sock_udp_send(sock, buf, pdu_len, NULL);
         if (res <= 0) {
@@ -157,27 +162,32 @@ static ssize_t _nanocoap_request(sock_udp_t *sock, coap_pkt_t *pkt, size_t len)
             break;
         }
 
-        res = sock_udp_recv(sock, buf, len, timeout, NULL);
-        if (res <= 0) {
-            if (res == -ETIMEDOUT) {
-                DEBUG("nanocoap: timeout\n");
-
-                timeout *= 2;
-                tries_left--;
-                if (!tries_left) {
-                    DEBUG("nanocoap: maximum retries reached\n");
+        while (block_timeout < xtimer_now_usec()) {
+            res = sock_udp_recv(sock, buf, len, timeout, NULL);
+            if (res <= 0) {
+                if (res == -ETIMEDOUT) {
+                    DEBUG("nanocoap: timeout\n");
+
+                    timeout *= 2;
+                    tries_left--;
+                    if (!tries_left) {
+                        DEBUG("nanocoap: maximum retries reached\n");
+                    }
+                    continue;
                 }
-                continue;
+                DEBUG("nanocoap: error receiving coap response, %d\n", (int)res);
+                break;
             }
-            DEBUG("nanocoap: error receiving coap response, %d\n", (int)res);
-            break;
-        }
-        else {
-            if (coap_parse(pkt, (uint8_t *)buf, res) < 0) {
-                DEBUG("nanocoap: error parsing packet\n");
-                res = -EBADMSG;
+            else {
+                if (coap_parse(pkt, (uint8_t *)buf, res) < 0) {
+                    DEBUG("nanocoap: error parsing packet\n");
+                    res = -EBADMSG;
+                }
+                if (coap_get_id(pkt) != id) {
+                    res = -1;
+                }
+                break;
             }
-            break;
         }
     }

```

</details>

Reply:
@fjmolinas do you have a way to trigger this?

Reply:
while (block_timeout < xtimer_now_usec())

This'll probably fail every ~71 minutes. I've tried to factor out the timeout handling, mind to take a look?

Reply:
> @fjmolinas do you have a way to trigger this?

I was seeing the issue from time to time when setting up a wireless setup for the update or when there where more threads running, I'll try to get a clean way to reproduce.

Otherwise the changes look good.

Reply:
Actually I have an issue with this, as it is every time it gets a wrong block it will ask for the correct one again, which will only hinder the situation for the next block that is fetched.

I really think it should not call `sock_udp_send(sock, buf, pdu_len, NULL)` if its the wrong block.

Reply:
> I really think it should not call `sock_udp_send(sock, buf, pdu_len, NULL)` if its the wrong block.

Ups, you're right, that was not intended. I changed the case to do "res = 0".
While we're at it, should an invalid packet trigger a resend?

Reply:
> Ups, you're right, that was not intended. I changed the case to do "res = 0".

I would rather do `res = EBADMSG` or `ENOMSG`, since `0` should be the everything OK return code so it makes the code confusing.

Also `else if (coap_get_id(pkt) != id) {` instead of ` if (coap_get_id(pkt) != id) {`, ot need to check the id if the packet is malformed.

> While we're at it, should an invalid packet trigger a resend?

I would say no.

Reply:
Tested and it solves the issue :)